src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp

```diff
@@ -322,8 +322,13 @@ inline oop ShenandoahHeap::evacuate_object(oop p, Thread* thread) {
 
   // Try to install the new forwarding pointer.
   oop copy_val = cast_to_oop(copy);
-  ContinuationGCSupport::relativize_stack_chunk(copy_val);
-
+  if (!copy_val->mark().is_marked()) {
+    // If we copied a mark-word that indicates 'forwarded' state, then
```

Comment:
Ouch. This is only the problem with `UseCompactObjectHeaders`, right? Can additionally conditionalize on that, so that legacy code path stays the same.

Reply:
Yes, upstreaming this separately would be cleaner for this PR. It would also highlight any problems with adjusting the lifecycle for relativization of stack chunks.